internal/app/referenceclient/transports.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2023-2024 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package referenceclient
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"connectrpc.com/conformance/internal"
+	conformancev1 "connectrpc.com/conformance/internal/gen/proto/go/connectrpc/conformance/v1"
+	"connectrpc.com/conformance/internal/gen/proto/go/connectrpc/conformance/v1/conformancev1connect"
+	"github.com/quic-go/quic-go"
+	"github.com/quic-go/quic-go/http3"
+)
+
+type transportSpec struct {
+	httpVersion   conformancev1.HTTPVersion
+	serverTLSCert string
+	clientTLSCert string
+	clientTLSKey  string
+}
+
+type transports struct {
+	cache sync.Map // map[transportSpec]http.RoundTripper
+}
+
+func (t *transports) get(req *conformancev1.ClientCompatRequest) (http.RoundTripper, *url.URL, error) {
+	tlsConf, err := createTLSConfig(req)
+	if err != nil {
+		return nil, nil, err
+	}
+	var scheme string
+	if tlsConf != nil {
+		scheme = "https://"
+	} else {
+		scheme = "http://"
+	}
+	urlString := scheme + net.JoinHostPort(req.Host, strconv.Itoa(int(req.Port)))
+	serverURL, err := url.ParseRequestURI(urlString)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid url: %s", urlString)
```

Comment:
While moving this, I noticed the previous `errors.New` with a format string was probably not intended

Reply:
Thanks for fixing!